code/modules/shuttle/computer.dm

```diff
@@ -5,69 +5,108 @@
-			if(admin_controlled)
-				dat += "Authorized personnel only<br>"
-				dat += "<A href='?src=[REF(src)];request=1]'>Request Authorization</A><br>"
-	dat += "<a href='?src=[REF(user)];mach_close=computer'>Close</a>"
-
-	var/datum/browser/popup = new(user, "computer", M ? M.name : "shuttle", 300, 200)
-	popup.set_content("<center>[dat]</center>")
-	popup.open()
+	data["docked_location"] = M ? M.get_status_text_tgui() : "Unknown"
+	data["status"] = M.mode == SHUTTLE_IGNITING ? "Igniting" : M.mode != SHUTTLE_IDLE ? "In Transit" : "Idle"
+	data["locations"] = list()
+	data["locked"] = FALSE
+	data["authorization_required"] = admin_controlled
+	data["timer_str"] = M ? M.getTimerStr() : "00:00"
+	data["destination"] = destination
+	if(admin_controlled)
+		data["status"] = "Unauthorized Access"
+	if(!M)
+		data["status"] = "Missing"
+		return data
+	for(var/obj/docking_port/stationary/S in SSshuttle.stationary)
+		if(!options.Find(S.id))
+			continue
+		if(!M.check_dock(S, silent = TRUE))
+			continue
+		var/list/location_data = list(
+			id = S.id,
+			name = S.name
+		)
+		data["locations"] += list(location_data)
+	if(length(data["locations"]) == 1)
+		for(var/location in data["locations"])
+			destination = location["id"]
+			data["destination"] = destination
+	if(!length(data["locations"]))
+		data["locked"] = TRUE
+		data["status"] = "Locked"
+	return data
 
-/obj/machinery/computer/shuttle/Topic(href, href_list)
-	if(..())
+/obj/machinery/computer/shuttle/ui_act(action, params)
+	. = ..()
+	if(.)
 		return
-	usr.set_machine(src)
-	src.add_fingerprint(usr)
 	if(!allowed(usr))
 		to_chat(usr, "<span class='danger'>Access denied.</span>")
 		return
 
-	if(href_list["move"])
-		var/obj/docking_port/mobile/M = SSshuttle.getShuttle(shuttleId)
-		if(M.launch_status == ENDGAME_LAUNCHED)
-			to_chat(usr, "<span class='warning'>You've already escaped. Never going back to that place again!</span>")
-			return
-		if(no_destination_swap)
-			if(M.mode == SHUTTLE_RECHARGING)
-				to_chat(usr, "<span class='warning'>Shuttle engines are not ready for use.</span>")
+	switch(action)
+		if("move")
+			var/obj/docking_port/mobile/M = SSshuttle.getShuttle(shuttleId)
+			if(M.launch_status == ENDGAME_LAUNCHED)
+				to_chat(usr, "<span class='warning'>You've already escaped. Never going back to that place again!</span>")
+				return
+			if(no_destination_swap)
+				if(M.mode == SHUTTLE_RECHARGING)
+					to_chat(usr, "<span class='warning'>Shuttle engines are not ready for use.</span>")
+					return
+				if(M.mode != SHUTTLE_IDLE)
+					to_chat(usr, "<span class='warning'>Shuttle already in transit.</span>")
+					return
+			var/list/options = params2list(possible_destinations)
+			if(!(params["shuttle_id"] in options))
+				log_admin("[usr] attempted to href dock exploit on [src] with target location \"[params["shuttle_id"]]\"")
+				message_admins("[usr] just attempted to href dock exploit on [src] with target location \"[params["shuttle_id"]]\"")
```

Comment:
can href exploits even happen on tgui?

Reply:
It's pretty easy to fake the contents of a tgui act

Reply:
Also reminder that this’ll catch no one because people codedive to make sure that hrefs work in the first place

Reply:
Just a legacy sanity check, I don't personally know how would they even try to do this. I don't expect it to catch anyone either, but should it happen, its probably best to have it logged.